<a href="https://colab.research.google.com/github/mayleebrewer/Discrete_Math/blob/main/BrewerMaylee_3913_Lesson6_Numbers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Project Part #6: CPSMA 3913 Dr. Lynam's Discrete Mathematics 

Maylee Brewer 262854

Numbers 

#1. Compute using modulo arithmetic

a) Convert the number $24036_7$ into base ten


I wanted to know what it be in other bases too so I tried in binary but it would not work. It did work for base 8 though. 

In [7]:
int('24036',7)

6201

b) Convert the hexadecimal $72e3$ into base ten

In [10]:
int('72e3',16)

29411

c) Convert 3735928559 into hexadecimal

In [2]:
hex(3735928559)

'0xdeadbeef'

Dead beef? Okay...

d) Using the two's complement preform the base ten subtraction $25-37$

All right let's convert these to binary

In [4]:
bin(25)

'0b11001'

In [7]:
bin(37)

'0b100101'

So since 37 is negitive we have to add one to the reverse. Then we would get 011010+1=101101. Now for some code

In [13]:
from math import floor
def binaryconvertor(x):
  ans = ''
  while x != 0:
    ans = str(x%2) + ans
    x = floor(x/2)
  return ans
binaryconvertor(25)

'11001'

In [16]:
def binaryconvertor(x):
  ans = ''
  if x<0:
    x = abs(x)
    ans = 'neg'
  while x != 0:
    ans = str(x%2) + ans
    x = floor(x/2)
  return ans
binaryconvertor(-37)

'100101neg'

In [17]:
def addone(list7binary):
  counter = 7
  list7binary[counter] = str(1 + int(list7binary[counter]))
  while list7binary[counter]=='2':
    list7binary[counter] = '0'
    counter -= 1
    list7binary[counter] = str(1 + int(list7binary[counter]))
  return list7binary

addone(['0','1','1','1','1','1','1','1'])

['1', '0', '0', '0', '0', '0', '0', '0']

Now we'll do the reverse. 

In [18]:
def flipit(list7binary):
  newlist = []
  for i in list7binary:
    if i == '0':
      newlist.append('1')
    else:
      newlist.append('0')
  return newlist

flipit(['0','1','1','1','1','1','1','1'])

['1', '0', '0', '0', '0', '0', '0', '0']

In [20]:

def twoscomplement(x):
  if not (x>-128)&(x<127):
    return "No Two's Complement"
  ans = '00000000'
  ans = list(ans) #strings don't allow you to change them in place (immutable) but lists can be!
  counter = 7
  if x>=0:
    while x != 0:
      ans[counter] = str(x%2)
      x = floor(x/2)
      counter -= 1
  if x<0:
    x = abs(x)
    while x != 0:
      ans[counter] = str(x%2)
      x = floor(x/2)
      counter -= 1
    ans = flipit(ans)
    addone(ans)        
  ans = ''.join(ans)
  return ans


In [21]:
twoscomplement(-37)

'11011011'

Now we have $11001+11011011= 1110100$,and I know that 25-37 is -12 so I will test it and make sure 

In [26]:
twoscomplement(-12)

'11110100'

I think it worked!

I think that these different bases are important to computers because in binary you only have two options and that helps computers to be faster but in hexadecimal it can condense big numbers by throwing some letters in. 

